File: RotateList.java
/**
 * Created by yhf on 15/5/24.
 */

/*
Given a list, rotate the list to the right by k places, where k is non-negative.

For example:
Given 1->2->3->4->5->NULL and k = 2,
return 4->5->1->2->3->NULL.
*/

public class RotateList {
    public static ListNode rotateRight(ListNode head, int k) {
        if (head == null || head.next == null) {
            return head;
        }
        int len = 0;
        ListNode p = head, fast = head, slow = head;
        while (p != null) {
            p = p.next;
            len++;
        }
        k = k % len;
        if (k == 0) {      // 注意加这一段
            return head;
        }
        while (k != 0) {
            fast = fast.next;
            k--;
        }
        while (fast.next != null) {
            fast = fast.next;
            slow = slow.next;
        }
        ListNode newHead = slow.next;
        slow.next = null;
        fast.next = head;
        return newHead;
    }

    public static void main(String[] args) {
        ListNode n1 = new ListNode(1);
        ListNode n2 = new ListNode(2);
        ListNode n3 = new ListNode(3);
        ListNode n4 = new ListNode(4);
        ListNode n5 = new ListNode(5);
        n1.next = n2;
        n2.next = n3;
        n3.next = n4;
        n4.next = n5;
        ListNode head = rotateRight(n1, 2);
    }
}

File: RotateList.py
"""
Given a list, rotate the list to the right by k places, where k is non-negative.

For example:
Given 1->2->3->4->5->NULL and k = 2,
return 4->5->1->2->3->NULL.
"""
from ListNode import ListNode


class Solution:
    # @param {ListNode} head
    # @param {integer} k
    # @return {ListNode}
    def rotateRight(self, head, k):
        temp, temp_k = head, k
        if not head or not head.next:
            return head
        while k > 0:
            head = self.rotateRightByOne(head)
            k -= 1
            if head == temp and temp_k != k:
                k %= (temp_k - k)
        return head

    def rotateRightByOne(self, head):
        cur = head
        while cur.next and cur.next.next:
            cur = cur.next
        new_head = cur.next
        cur.next = None
        new_head.next = head
        return new_head


if __name__ == '__main__':
    head = ListNode(1, 2, 3, 4)
    print Solution().rotateRight(head, 3)
